jobset-generator/jobsets-builder.py

```diff
@@ -0,0 +1,290 @@
+    inputs: JobInputCollection
+
+
+class HydraJobFlake(TypedDict):
+    flake: str
+
+
+HydraInputDefinition = Union[HydraJobLegacy, HydraJobFlake]
+
+
+class HydraJob(TypedDict):
+    enabled: bool
+    hidden: bool
+    description: str
+    checkinterval: int
+    schedulingshares: int
+    enableemail: bool
+    emailoverride: str
+    keepnr: int
+    definition: HydraInputDefinition
+
+
+class FlattenedHydraJob(TypedDict):
+    enabled: bool
+    hidden: bool
+    description: str
+    checkinterval: int
+    schedulingshares: int
+    enableemail: bool
+    emailoverride: str
+    keepnr: int
+    flake: Optional[str]
+    nixexprinput: Optional[str]
+    nixexprpath: Optional[str]
+    inputs: Optional[JobInputCollection]
+
+
+class JobConfig(TypedDict):
+    checkinterval: int
+    emailoverride: str
+    enableemail: bool
+    keepnr: int
+    schedulingshares: int
+    input_template: JobInputCollection
+    email_responsible: bool
+    inputname: str
+    inputpath: str
+
+
+def flatten_hydra_job(job: HydraJob) -> FlattenedHydraJob:
+    flat = FlattenedHydraJob(
+        enabled=job["enabled"],
+        hidden=job["hidden"],
+        description=job["description"],
+        checkinterval=job["checkinterval"],
+        schedulingshares=job["schedulingshares"],
+        enableemail=job["enableemail"],
+        emailoverride=job["emailoverride"],
+        keepnr=job["keepnr"],
+        flake=None,
+        nixexprinput=None,
+        nixexprpath=None,
+        inputs=None,
+    )
+
+    if "flake" in job["definition"]:
+        flake: HydraJobFlake = job["definition"]  # type: ignore
+        flat["flake"] = flake["flake"]
+    elif "nixexprinput" in job["definition"]:
+        legacy: HydraJobLegacy = job["definition"]  # type: ignore
+        flat["nixexprinput"] = legacy["nixexprinput"]
+        flat["nixexprpath"] = legacy["nixexprpath"]
+        flat["inputs"] = legacy["inputs"]
+
+    return flat
+
+
+MakeDefinition = Callable[[JobConfig, PullRequest], HydraInputDefinition]
+
+
+def make_flake_definition(
+    _job_config: JobConfig, pr: PullRequest
+) -> HydraInputDefinition:
+    return HydraJobFlake(
+        flake="git+ssh://{}?{}".format(
+            pr["head"]["repo"]["ssh_url"],
+            urllib.parse.urlencode(
+                {
+                    "ref": pr["head"]["ref"],
+                    "rev": pr["head"]["sha"],
+                }
+            ),
+        )
+    )
+
+
+def make_legacy_definition(
+    job_config: JobConfig, pr: PullRequest
+) -> HydraInputDefinition:
+    inputs: JobInputCollection = job_config["input_template"]
+
+    inputs["src"] = HydraJobInput(
+        type="git",
+        value="{} {}".format(pr["head"]["repo"]["git_url"], pr["head"]["sha"]),
+        emailresponsible=job_config["email_responsible"],
+    )
+
+    return HydraJobLegacy(
+        nixexprinput=job_config["inputname"],
+        nixexprpath=job_config["inputpath"],
+        inputs=inputs,
+    )
+
+
+def main(
```

Comment:
Nitpick-y, but I would swap `main` and `make_job` so that the main function is closer to the `if __name__` check.